pkgs/record_use/lib/src/data_classes/annotation.dart

```diff
@@ -0,0 +1,44 @@
+// Copyright (c) 2024, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:collection/collection.dart';
+
+import 'identifier.dart';
+
+class Annotation {
+  final Identifier identifier;
+  final Map<String, dynamic> fields;
+
+  Annotation({
+    required this.identifier,
+    required this.fields,
+  });
+
+  factory Annotation.fromJson(
+    Map<String, dynamic> json,
+    List<Identifier> identifiers,
+  ) =>
+      Annotation(
+        identifier: identifiers[json['id'] as int],
+        fields: json['fields'] as Map<String, dynamic>,
+      );
+
+  Map<String, dynamic> toJson(List<Identifier> identifiers) => {
+        'id': identifiers.indexOf(identifier),
+        'fields': fields,
+      };
+
+  @override
+  bool operator ==(Object other) {
+    if (identical(this, other)) return true;
+    final mapEquals = const DeepCollectionEquality().equals;
+
+    return other is Annotation &&
+        other.identifier == identifier &&
+        mapEquals(other.fields, fields);
+  }
+
+  @override
+  int get hashCode => identifier.hashCode ^ fields.hashCode;
```

Comment:
Using `^` doesn't scramble the bits very well, you can use `Object.hash(a, b)`.

Also `fields.hashCode` isn't based on the content of the hash map, instead it uses a random number assigned to this instance of a map. So it won't give the right result

If one wants content based comparison / hashing, then one has to use content in `get hashCode` and `operator==`


Do we really need this hashability?

It seems tedious and error prone (as we see with this bug) to maintain this code. Could this be generated (json serializable, built value, ...)?

Reply:
I wanted to avoid adding dependencies, and I am not a fan of generated code. But I see the point about this being error-prone. I added a dependency on `equatable`, which I think is pretty neat.